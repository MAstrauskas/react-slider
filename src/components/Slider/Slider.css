.slider {
  display: flex;
  flex-direction: column;
  border: 1px solid rgba(0, 0, 0, 0.3);
  /*width: 57.7%;*/
  max-width: 68em;
  background-color: #f6f6f6;
  overflow: hidden;
}

.slider-wrapper {
  position: relative;
  display: flex;
  flex: 1;
  align-self: baseline;
}

.slider-arrows {
  display: flex;
  width: 40px;
  align-self: center;
}

/* Responsiveness */

/* Notebooks & Tablets */
@media (max-width: 1100px) {
  .slider {
    max-width: 45em;
  }
}

/* Mobile phones */
@media (max-width: 750px) {
  .slider {
    max-width: 22em;
  }
}
