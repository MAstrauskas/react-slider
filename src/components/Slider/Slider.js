import React, { Component } from "react";
import Card from "../Card/Card";
import LeftArrow from "../Arrows/LeftArrow";
import RightArrow from "../Arrows/RightArrow";

import "./Slider.css";
import {ResizeObserver} from "resize-observer";

class Slider extends Component {
  constructor(props) {
    // Pass props to the Parent component
    super(props);

    // Set initial state
    this.state = {
      cards: [],
      currentIndex: 0,
      translateValue: 0,
      isError: false,
      slidesPerPage: 1,
    };

    this.updateScreenSizes = this.updateScreenSizes.bind(this);
  }

  componentDidMount() {
    this.fetchCard();
    this.updateScreenSizes(this.state.slidesPerPage);
    window.addEventListener("resize", () => {
      this.updateScreenSizes(this.state.slidesPerPage);
    });
  }

  componentWillUnmount() {
    window.removeEventListener("resize", () => {
      this.updateScreenSizes(this.state.slidesPerPage);
    });
  }

  fetchCard() {
    fetch("http://127.0.0.1:3001/cards/")
      .then(response => response.json())
      .then(cards => this.setState({ cards }))
      .catch(() => this.setState({ isError: true }));
  }

  updateScreenSizes(slidesNumber) {
    if (window.innerWidth > 1100) {
      slidesNumber = 3;
    } else if (window.innerWidth > 750) {
      slidesNumber = 2;
    } else if (window.innerWidth <= 750) {
      slidesNumber = 1;
    }

    this.setState({
      slidesPerPage: slidesNumber
    });
  }

  handlePrevSlide = () => {
    // Stay at the front of cards array if
    // currentIndex is 0, otherwise
    // move left by one card at a time
    if (this.state.currentIndex === 0) {
      return this.setState({
        currentIndex: 0,
        translateValue: 0
      });
    }

    this.setState(prevState => ({
      currentIndex: prevState.currentIndex - 1,
      translateValue: prevState.translateValue - -this.customWidth()
    }));
  };

  handleNextSlide = () => {
    const { slidesPerPage } = this.state;

    // If at the end of cards array, exits and
    // currentIndex and translateValue resets to 0
    // Also, according to the size of the screen
    // and the number of cards it
    // sets when to return to the first card.

      if (this.state.currentIndex === this.state.cards.length - slidesPerPage) {
        return this.setState({
          currentIndex: 0,
          translateValue: 0
        });
      }


    // Only run if the condition above
    // is not met.
    this.setState(prevState => ({
      currentIndex: prevState.currentIndex + 1,
      translateValue: prevState.translateValue + -this.customWidth()
    }));
  };

  customWidth = () => {
    return 360;
  };

  render() {
    const isError  = this.state.isError;

    if (isError) {
      return (
        <div>
          <h1>API Server down. Please try again later.</h1>
        </div>
      );
    }
    return (
      <div className="slider">
        <div
          className="slider-wrapper"
          style={{
            transform: `translateX(${this.state.translateValue}px)`,
            transition: "transform ease-out 0.5s"
          }}
        >
          {this.state.cards.map(item => (
            <Card
              key={item.id}
              id={item.id}
              title={item.title}
              subtitle={item.subtitle}
              text={item.text}
              image={item.image_url}
              href={item.href}
              is_liked={item.is_liked}
            />
          ))}
        </div>
        <div className="slider-arrows">
          <LeftArrow handlePrevSlide={this.handlePrevSlide} />
          <RightArrow handleNextSlide={this.handleNextSlide} />
        </div>
      </div>
    );
  }
}

export default Slider;
