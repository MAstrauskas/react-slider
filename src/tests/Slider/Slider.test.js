import React from "react";
import { create } from "react-test-renderer";
import Slider from "../../components/Slider/Slider";

describe("Slider Component", () => {
    test("it matches the snapshot", () => {
        const component = create(<Slider/>);
        expect(component.toJSON()).toMatchSnapshot();
    });
});