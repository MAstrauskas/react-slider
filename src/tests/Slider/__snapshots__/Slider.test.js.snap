// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Component it matches the snapshot 1`] = `
<div
  className="slider"
>
  <div
    className="slider-wrapper"
    style={
      Object {
        "transform": "translateX(0px)",
        "transition": "transform ease-out 0.5s",
      }
    }
  />
  <div
    className="slider-arrows"
  >
    <div
      className="arrow"
      onClick={[Function]}
    >
      <img
        alt="Left Arrow"
        aria-hidden="true"
        src="left-arrow.svg"
      />
    </div>
    <div
      className="arrow"
      onClick={[Function]}
    >
      <img
        alt="Right Arrow"
        aria-hidden="true"
        src="right-arrow.svg"
      />
    </div>
  </div>
</div>
`;
